src/main/java/org/folio/uk/integration/keycloak/KeycloakService.java

```diff
@@ -23,37 +23,113 @@
 import org.apache.commons.lang3.StringUtils;
 import org.folio.spring.FolioExecutionContext;
 import org.folio.uk.domain.dto.User;
-import org.folio.uk.exception.RequestValidationException;
+import org.folio.uk.domain.dto.UserTenant;
+import org.folio.uk.integration.keycloak.config.KeycloakFederatedAuthProperties;
 import org.folio.uk.integration.keycloak.config.KeycloakLoginClientProperties;
 import org.folio.uk.integration.keycloak.model.Client;
 import org.folio.uk.integration.keycloak.model.Credential;
+import org.folio.uk.integration.keycloak.model.FederatedIdentity;
 import org.folio.uk.integration.keycloak.model.KeycloakUser;
 import org.folio.uk.integration.keycloak.model.ScopePermission;
+import org.folio.uk.integration.users.UserTenantsClient;
 import org.springframework.stereotype.Component;
 
 @Log4j2
 @Component
 @RequiredArgsConstructor
 public class KeycloakService {
 
+  private static final int RANDOM_STRING_COUNT = 6;
   private final KeycloakClient keycloakClient;
   private final TokenService tokenService;
+  private final UserTenantsClient userTenantsClient;
   private final FolioExecutionContext folioExecutionContext;
   private final KeycloakLoginClientProperties loginClientProperties;
+  private final KeycloakFederatedAuthProperties keycloakFederatedAuthProperties;
 
-  public void createUser(User user, String password) {
-    if (user.getId() == null) {
-      throw new RequestValidationException("User id is missing", "id", user.getId());
+  public String createUser(User user, String password) {
+    var foundKcUser = findUserByUsername(user.getUsername(), false);
+    if (foundKcUser.isPresent()) {
+      var kcUserId = foundKcUser.get().getId();
+      var kcUserAttributes = foundKcUser.flatMap(this::getUserIdFromAttributes);
+      kcUserAttributes.ifPresent(userId -> updateUser(fromString(userId), user));
+
+      return kcUserId;
     }
+
     var kcUser = toKeycloakUser(user, password);
     kcUser.setUserTenantAttr(of(getRealm()));
     log.info("Creating keycloak user: userId = {}", user.getId());
+    return callKeycloak(
+      create(kcUser), () -> buildUsersErrorMessage("Failed to create keycloak user", user.getId()));
+  }
+
+  public void linkIdentityProviderToUser(User user, String kcUserId) {
+    var userId = user.getId();
+    var tenant = getRealm();
+
+    log.info("Linking identity provider to user [userId: {}, kcUserId: {}, tenant: {}, username: {}]", userId,
+      kcUserId, tenant, user.getUsername());
```

Comment:
storing username in logs can help identify user identity to understand what was performed with particular user and so violate PII, so probably lets avoid logging it